tka/deeplink.go

```diff
@@ -0,0 +1,162 @@
+// The input urlString follows this structure:
+//
+// tailscale://sign-device/v1/?nk=xxx&tp=xxx&dn=xxx&os=xxx&em=xxx&hm=xxx
+//
+// where:
+// - "nk" is the nodekey of the node being signed
+// - "tp" is the tailnet lock public key
+// - "dn" is the name of the node
+// - "os" is the operating system of the node
+// - "em" is the email address associated with the node
+// - "hm" is a SHA-256 HMAC computed over the concatenation of the above fields
+func (a *Authority) ValidateDeeplink(urlString string) DeeplinkValidationResult {
+	parsedUrl, err := url.Parse(urlString)
+	if err != nil {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   err,
+		}
+	}
+
+	if parsedUrl.Scheme != DeeplinkTailscaleURLScheme {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("unhandled scheme %s, expected %s", parsedUrl.Scheme, DeeplinkTailscaleURLScheme),
+		}
+	}
+
+	if parsedUrl.Host != DeeplinkCommandSign {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("unhandled host %s, expected %s", parsedUrl.Host, DeeplinkCommandSign),
+		}
+	}
+
+	path := parsedUrl.EscapedPath()
+	pathComponents := strings.Split(path, "/")
+	if len(pathComponents) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("no path components found"),
+		}
+	}
+
+	if pathComponents[0] != "v1" {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("expected v1 deeplink version, found something else"),
+		}
+	}
+
+	nodeKey := parsedUrl.Query().Get("nk")
+	if len(nodeKey) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("missing nk (NodeKey) query parameter"),
+		}
+	}
+
+	tlPub := parsedUrl.Query().Get("tp")
+	if len(tlPub) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("missing tp (TLPub) query parameter"),
+		}
+	}
+
+	deviceName := parsedUrl.Query().Get("dn")
+	if len(deviceName) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("missing dn (DeviceName) query parameter"),
+		}
+	}
+
+	osName := parsedUrl.Query().Get("os")
+	if len(deviceName) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("missing os (OSName) query parameter"),
+		}
+	}
+
+	emailAddress := parsedUrl.Query().Get("em")
+	if len(emailAddress) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("missing em (EmailAddress) query parameter"),
+		}
+	}
+
+	hmacString := parsedUrl.Query().Get("hm")
+	if len(hmacString) == 0 {
+		return DeeplinkValidationResult{
+			IsValid: false,
+			Error:   fmt.Errorf("missing hm (HMAC) query parameter"),
+		}
+	}
+
+	components := []string{nodeKey, tlPub, deviceName, osName, emailAddress}
+	stateID1, _ := a.StateIDs()
+	computedHMAC := generateHMAC(stateID1, components)
+
+	if computedHMAC != hmacString {
```

Comment:
Not sure how relevant it is here, but the `hmac` package has an explicit `Equals` function which doesn't leak timing information. Might be a better practice to have `generateHMAC` return a `[]byte` and then use:

```go
if !hmac.Equals(computedHMAC, hmacString) {
  return DeeplinkValidationResult{...}
}
```